nikola/post.py

```diff
@@ -934,129 +942,103 @@ def source_ext(self, prefix=False):
+                break
+        return meta, used_extractor
     except (UnicodeDecodeError, UnicodeEncodeError):
         msg = 'Error reading {0}: Nikola only supports UTF-8 files'.format(source_path)
         LOGGER.error(msg)
         raise ValueError(msg)
     except Exception:  # The file may not exist, for multilingual sites
-        return {}
+        return {}, None
 
 
-re_md_title = re.compile(r'^{0}([^{0}].*)'.format(re.escape('#')))
-# Assuming rst titles are going to be at least 4 chars long
-# otherwise this detects things like ''' wich breaks other markups.
-re_rst_title = re.compile(r'^([{0}]{{4,}})'.format(re.escape(
-    string.punctuation)))
-
-
-def _get_metadata_from_file(file_lines, config=None):
-    """Extract metadata from a post's source file."""
-    meta, metadata_type = utils.extract_metadata(file_lines)
-    if metadata_type in ('toml', 'yaml'):
-        # Map metadata from other platforms to names Nikola expects (Issue #2817)
-        map_metadata(meta, metadata_type, config)
-    return meta
-
-
-def get_metadata_from_meta_file(path, config=None, lang=None):
+def get_metadata_from_meta_file(path, post, config, lang, metadata_extractors_by=None):
     """Take a post path, and gets data from a matching .meta file."""
     meta_path = os.path.splitext(path)[0] + '.meta'
     if lang and config:
         meta_path = get_translation_candidate(config, meta_path, lang)
     elif lang:
         meta_path += '.' + lang
     if os.path.isfile(meta_path):
-        return get_metadata_from_file(path, config, lang)
-
+        return get_metadata_from_file(meta_path, post, config, lang, metadata_extractors_by)[0]
     elif lang:
         # Metadata file doesn't exist, but not default language,
         # So, if default language metadata exists, return that.
         # This makes the 2-file format detection more reliable (Issue #525)
-        return get_metadata_from_meta_file(path, config, lang=None)
+        return get_metadata_from_meta_file(meta_path, post, config, None, metadata_extractors_by)
     else:  # No 2-file metadata
         return {}
 
 
-def get_meta(post, file_metadata_regexp=None, unslugify_titles=False, lang=None):
-    """Get post's meta from source.
-
-    If ``file_metadata_regexp`` is given it will be tried to read
-    metadata from the filename.
-    If ``unslugify_titles`` is True, the extracted title (if any) will be unslugified, as is
-    done in galleries. If any metadata is then found inside the file the metadata from the
-    file will override previous findings.
-    """
+def get_meta(post, lang):
+    """Get post meta from compiler or source file."""
     meta = defaultdict(lambda: '')
+    used_extractor = None
 
-    try:
-        config = post.config
-    except AttributeError:
-        config = None
+    config = getattr(post, 'config', None)
+    metadata_extractors_by = getattr(post, 'metadata_extractors_by', metadata_extractors.default_metadata_extractors_by())
```

Comment:
This always calls `metadata_extractors.default_metadata_extractors_by()`, even though it should never be used. Wouldn't it be better to add two more lines (`if metadata_extractors_by is None:` / `metadata_extractors_by = metadata_extractors.default_metadata_extractors_by()`) to avoid this call every time?